Add progress reporting for imports

daviddavis opened the pull request:
fixes #6559

Please be sure you have read our documentation on creating PRs:
https://docs.pulpproject.org/contributing/pull-request-walkthrough.html

goosemania commented on pulpcore/app/tasks/importer.py:
```diff
@@ -178,6 +182,20 @@ def import_repository_version(importer_pk, destination_repo_pk, source_repo_name
+        content_count = content.count()
+        data = dict(
+            message=f"Importing content for {dest_repo.name}",
+            code="import-repo-version-content",
```
I think we have an unspoken and unwritten convention :) to use dots in the `code` as a word separator. I don't have any preference, just for the sake of consistency.

ggainey replied:
Well, darn, I did not realize this - pulpcore/app/importexport.py uses dashes as well :(

daviddavis replied:
yea, that's where I got it from.

goosemania commented on pulpcore/app/tasks/importer.py:
```diff
@@ -337,18 +358,27 @@ def validate_and_assemble(toc_filename):
+        data = dict(message="Importing Artifacts", code="import-artifacts",
+                    total=len(ar_result.rows))
+        with ProgressReport(**data) as pb:
+            for row in pb.iter(ar_result.rows):
```
It's user friendly to show progress incrementing progress report counters for each artifact however it might affect performance I think. 
The whole operation is not fast anyway , so maybe it's not a concern

daviddavis replied:
I thought that's what this code is doing?

goosemania replied:
It is. Just shared that performance might be a potential concern here.

ggainey replied:
At this step we're physically moving Artifact binaries to their final resting-places on the filesystem; I expect the I/O time will completely swamp the time to update the progress report. SSDs are fast, but it's a LOT of bytes :)

dralley replied:
We really need to look into async file I/O

dralley replied:
I guess technically threads are fine for this purpose as well.  If it becomes a thing that people complain about being slow anyways.

goosemania commented on pulpcore/app/tasks/importer.py:
```diff
@@ -337,18 +358,27 @@ def validate_and_assemble(toc_filename):
+            gpr = GroupProgressReport(
+                message="Importing repository versions",
+                code="import-repo-versions",
+                total=len(data),
```
Maybe we should add `done` here as well. What do you think?
The default is 0 https://github.com/pulp/pulpcore/blob/master/pulpcore/app/models/progress.py#L276

ggainey commented:
After running CentOS8-base through the export/import dance, I have a couple of observations from the end-user's POV. 

With a large export-file, it can take an appreciable time (ones of minutes) before we get any progress reporting in the parent task or task-group. Similarly, once we dispatch the child-tasks-per-repo, there's a lot of startup going on before we have anything in the task-progress-reports. I'm not sure that this kind of reporting fits into the progress-report framework. As it stands, to really see what's going on, the user wants to be watching the systemctl logs in addition to the progress-reports.

Maybe we just need to document this expectation somewhere?

daviddavis commented:
I think I addressed all the feedback. 

I also added some additional progress reporting and updated some of the existing progress reporting. Namely, I added progress reports to validating and recombining the chunks so progress reporting starts instantly when the task starts. Also, I create the import version progress report at the beginning of the child task code instead of at the end so there is a progress report immediately available. 

It's not perfect but I think that should solve the gaps between when the tasks start and when progress reporting starts.

goosemania commented:
Binding tests fail because fedorapeople was unavailable and they reference old fixture repos:
```
$ git grep repos.fedora
.ci/assets/bindings/test_bindings.py:        "https://repos.fedorapeople.org/repos/pulp/pulp/demo_repos/test_bandwidth_repo/"
.ci/assets/bindings/test_bindings.py:remote_url = "https://repos.fedorapeople.org/pulp/pulp/demo_repos/test_file_repo/PULP_MANIFEST"
.ci/assets/bindings/test_bindings.rb:remote_url = 'https://repos.fedorapeople.org/pulp/pulp/demo_repos/test_file_repo/PULP_MANIFEST'
```
Do we need move those to the fixtures.pulpproject.org or they didn't migrate for a reason?

goosemania reviewed (approved):
Thank you, @daviddavis !

daviddavis commented:
fixtures.pulpproject.org is a deployment of the pulp-fixtures repo (formerly at https://repos.fedorapeople.org/repos/pulp/pulp/fixtures/) and the demo repos were created outside of pulp-fixtures. We could move them into pulp-fixtures or onto the pulpproject.org server but I haven't done so because of how infrequently we use them in Pulp 3.

ggainey commented:
> It's not perfect but I think that should solve the gaps between when the tasks start and when progress reporting starts.

Doing some testing right now, and def concur - you've answered my user-experience issues completely.
